### PR TITLE
#33 Registration form (v2)

### PR DESCRIPTION
**Pull request by alinaah**

### CODE REVIEW

File: web/frontend/angular/src/app/main-pages/register/register.component.ts [hunk 1]
```diff
@@ -4,6 +4,7 @@ import {Validators} from "@angular/forms"
 import {FormBuilder} from "@angular/forms"
 import {Router} from "@angular/router"
 import {AuthService} from "src/app/auth/auth.service"
+import * as $ from "jquery"
```
Comment (MarkKoz) [inline web/frontend/angular/src/app/main-pages/register/register.component.ts:7]: You never installed jQuery, so this import fails.

Honestly, I'm not a fan of jQuery. I believe modern JavaScript is perfectly capable of achieving the same things in most cases (albeit with a more verbose syntax sometimes). jQuery feels like bloat to me. Is it even beneficial given that we're using Angular as a framework? Does Angular not have a better way of doing this? Is it possible for you to rewrite your error handling code without using jQuery?

File: web/settings.py [hunk 2]
```diff
@@ -81,7 +81,7 @@
 
 REST_FRAMEWORK = {
     "DEFAULT_AUTHENTICATION_CLASSES": [
-        "rest_framework.authentication.SessionAuthentication",
+        "rest_framework.authentication.BasicAuthentication",
```
Comment (MarkKoz) [inline web/settings.py:84]: Why was this changed? I don't believe this is necessary, and basic authentication is generally only appropriate for testing purposes. Furthermore, django-rest-registration is designed around session/token based authentication.

File: web/frontend/angular/src/app/main-pages/register/register.component.ts [hunk 3]
```diff
@@ -36,7 +38,20 @@ export class RegisterComponent implements OnInit {
...
+                this.router.navigateByUrl("homepage")
+            },
+            (error) => {
+                console.log(error)
+                $("input[type=email],input[type=password]").addClass("error")
+                $("#error").removeClass("d-none")
+                setTimeout(() => {
+                    $("input").removeClass("error")
+                    $("#error").addClass("d-none")
+                }, 2000)
```
Comment (MarkKoz) [inline web/frontend/angular/src/app/main-pages/register/register.component.ts:53]: It's better to leave the error visible indefinitely to ensure the user does not miss it. The error should go away if the user successfully registers.

File: web/frontend/angular/src/app/main-pages/register/register.component.ts [hunk 4]
```diff
@@ -24,6 +25,7 @@ export class RegisterComponent implements OnInit {
         this.authForm = this.formBuilder.group({
             email: ["", Validators.required],
```
Comment (MarkKoz) [inline web/frontend/angular/src/app/main-pages/register/register.component.ts:26]: An email is not required. It is optional. Only the username, password, and password confirmation are required. Does this form even have a username field?

File: web/frontend/angular/src/app/auth/auth.service.ts [hunk 5]
```diff
@@ -15,6 +17,11 @@ export class AuthService {
         localStorage.removeItem("ACCESS_TOKEN")
     }
     public register(userData: User) {
-        //TODO
+        userData.username = userData.email
```
Comment (MarkKoz) [inline web/frontend/angular/src/app/auth/auth.service.ts:20]: The username and the email are distinctly separate fields. It is not appropriate to make the email the username.

**Review by MarkKoz — CHANGES_REQUESTED** [review]

File: web/frontend/angular/src/app/main-pages/register/register.component.ts (hunk 3 above)
Comment (MarkKoz) [reply web/frontend/angular/src/app/main-pages/register/register.component.ts:53]: @alinaah You did not address this comment.

File: web/settings.py (hunk 2 above)
Comment (MarkKoz) [reply web/settings.py:84]: @alinaah You did not address this comment.

File: web/frontend/angular/src/app/main-pages/register/register.component.ts (hunk 1 above)
Comment (MarkKoz) [reply web/frontend/angular/src/app/main-pages/register/register.component.ts:7]: @alinaah You did not address this comment.

File: web/frontend/angular/src/app/main-pages/register/register.component.ts (hunk 4 above)
Comment (MarkKoz) [reply web/frontend/angular/src/app/main-pages/register/register.component.ts]: @alinaah I see you replaced the email with a username field. However, the form should have _both_ and email field and a username field. To be clear, the form should have the following fields:

* username (required)
* email (optional)
* password (required)
* confirm password (required)

File: web/frontend/angular/src/app/main-pages/register/register.component.ts (hunk 1 above)
Comment (MarkKoz) [reply web/frontend/angular/src/app/main-pages/register/register.component.ts:7]: Here are some alternatives to using jQuery for this. Please do further research on them, since I just gathered these from a quick search without really reading through them.

* https://www.npmjs.com/package/ngx-form-validations
* https://www.pluralsight.com/guides/how-to-display-validation-messages-using-angular
* https://stackoverflow.com/questions/52771445/best-way-to-show-error-messages-for-angular-reactive-forms-one-formcontrol-mult
* https://netbasal.com/make-your-angular-forms-error-messages-magically-appear-1e32350b7fa5

Comment (MarkKoz) [reply web/frontend/angular/src/app/main-pages/register/register.component.ts:7]: Since we will have many forms elsewhere, it'd be good to make this error handling re-usable in some way across all forms.

File: web/frontend/angular/src/app/main-pages/register/register.component.html [hunk 6]
```diff
@@ -5,12 +5,30 @@ <h2 class="register-header">Register</h2>
     class="register-container"
     (ngSubmit)="register()"
   >
-    <p [ngClass]="{'has-error': isSubmitted && formControls.email.errors}">
-      <input type="email" placeholder="Email" formControlName="email" />
+    <div class="row d-none" id="error">
+      <span>Invalid Credentials</span>
```
Comment (MarkKoz) [inline web/frontend/angular/src/app/main-pages/register/register.component.html:9]: You need to be dynamically changing the error message to correspond to the actual error that occurred. "Invalid credentials" is not always an appropriate error message.

Here is an example error that DRF may return
```json
{"username":["A user with that username already exists."]}
```

We use the default error configuration for DRF, so you can look at the documentation to get an understanding of the error response formats https://www.django-rest-framework.org/api-guide/exceptions/ Notice that validation errors (like the above example) are handled differently by DRF.

That being said, not _all_ errors will follow this format, because sometimes there may be an internal server error (500+ status code). In that case just display "internal error" as the message or something like that.

File: web/frontend/angular/src/app/main-pages/register/register.component.ts [hunk 7]
```diff
@@ -36,7 +38,20 @@ export class RegisterComponent implements OnInit {
...
             return
         }
-        this.authService.register(this.authForm.value)
-        this.router.navigateByUrl("homepage")
+        this.authService.register(this.authForm.value).subscribe(
+            (res: any) => {
+                console.log(res)
+                this.router.navigateByUrl("homepage")
+            },
+            (error) => {
```
Comment (MarkKoz) [inline web/frontend/angular/src/app/main-pages/register/register.component.ts:46]: You need to clear all the error stuff if a request succeeds. Currently, if the first request fails but the second succeeds, everything still appears red.

There should be some feedback to the user about successful registration. Like a message somewhere that says registration was successful.

File: web/settings.py (hunk 2 above)
Comment (MarkKoz) [reply web/settings.py:84]: This still has not been addressed.

File: web/frontend/angular/src/app/auth/auth.service.ts [hunk 8]
```diff
@@ -15,6 +17,10 @@ export class AuthService {
         localStorage.removeItem("ACCESS_TOKEN")
     }
     public register(userData: User) {
-        //TODO
+        console.log(userData)
```
Comment (MarkKoz) [inline web/frontend/angular/src/app/auth/auth.service.ts:20]: Remove this log. If needed, this data can be seen in the developer console by inspecting the network request in the network tab.

File: web/frontend/angular/src/app/main-pages/register/register.component.ts [hunk 9]
```diff
@@ -36,7 +38,20 @@ export class RegisterComponent implements OnInit {
         if (this.authForm.invalid) {
             return
         }
-        this.authService.register(this.authForm.value)
-        this.router.navigateByUrl("homepage")
+        this.authService.register(this.authForm.value).subscribe(
+            (res: any) => {
+                console.log(res)
```
Comment (MarkKoz) [inline web/frontend/angular/src/app/main-pages/register/register.component.ts:43]: Remove this log. If needed, this data can be seen in the developer console by inspecting the network request in the network tab.

File: web/frontend/angular/src/app/main-pages/register/register.component.ts [hunk 10]
```diff
@@ -36,7 +38,20 @@ export class RegisterComponent implements OnInit {
...
+                this.router.navigateByUrl("homepage")
+            },
+            (error) => {
+                console.log(error)
```
Comment (MarkKoz) [inline web/frontend/angular/src/app/main-pages/register/register.component.ts:47]: Remove this log. If needed, this data can be seen in the developer console by inspecting the network request in the network tab.

File: web/settings.py (hunk 2 above)
Comment (alinaah) [reply web/settings.py:84]: I can't do anything on this, otherwise Apis won't work. 
It is better to use token authentication. It is not a common practis of uning session authentication with apis. Session authentication do not work with apis, instead we should use token authentication.

Comment (MarkKoz) [reply web/settings.py:84]: > It is better to use token authentication.

No, in this case it is not. Token authentication makes sense for non-browser clients, but our client is the browser. Furthermore, basic authentication, which you have configured, is _not_ token authentication.

However, if you specifically meant a JSON Web Token, then it's a different story. But it'd require more work to set up, and regardless, I am doubtful of the benefits it provides in our case over session auth.

> It is not a common practis of uning session authentication with apis.

It is common and is the best choice when the client of the API is a browser.

> I can't do anything on this, otherwise Apis won't work.
> Session authentication do not work with apis

Yes, they do work. I have tested it and even demonstrated it in class during a presentation. I also left several comments on your previous PR with examples of how to authenticate using sessions by getting a CSRF token from a cookie and passing it as a header.

Comment (Ananthu17) [reply web/settings.py:84]: In order to do Session authentication i need to add csrf tokens. If it was  normal django template in frontend  then i can use {{csrf_token}}. But now the case is different we are using angular. {{csrf_token}} in <form> tag is not gonna work. You need to pass csrf token from backend when a user logged in. But i could n't find any.

File: web/frontend/angular/src/app/main-pages/register/register.component.html (hunk 6 above)
Comment (alinaah) [reply web/frontend/angular/src/app/main-pages/register/register.component.html:9]: Not every time Api response is like this. On postman i am getting response like this. But on Angular iam getting errror like this."HttpErrorResponse {headers: HttpHeaders, status: 400, statusText: 'Bad Request', url: 'http://0.0.0.0:8000/api/accounts/register/', ok: false, …}"

File: web/settings.py (hunk 2 above)
Comment (MarkKoz) [reply web/settings.py:84]: Yes all that is taken care of. I left a comment on your old PR which showed how to do this https://github.com/Comp-490-SeniorProject/site/pull/43#issuecomment-1045227851

Please note that you may need to adjust the code, as that was just an example in pure JS. Angular may have its own utilities for managing cookies, and of course its own way to make http requests (which you're already using in this PR).

File: web/frontend/angular/src/app/main-pages/register/register.component.html (hunk 6 above)
Comment (MarkKoz) [reply web/frontend/angular/src/app/main-pages/register/register.component.html]: According to documentation for [`HttpErrorResponse`](https://angular.io/api/common/http/HttpErrorResponse#description=),

> The error property will contain either a wrapped Error object or the error response returned from the server.

So, check the `error` property of that response.

File: web/frontend/angular/src/environments/environment.ts [hunk 11]
```diff
@@ -4,6 +4,7 @@
 
 export const environment = {
     production: false,
+    baseUrl: "http://0.0.0.0:8000/",
```
Comment (MarkKoz) [inline web/frontend/angular/src/environments/environment.ts:7]: This url does not work for me. Has it been working for you? It actually wouldn't make sense if it _does_ work.

Even the simple case of visiting `http://0.0.0.0:8000` in my browser does not work, nor would I expect it to. Even if it somehow did work, our security policy would reject it because it considers it to be a different domain:

> zone.js:2863 Refused to connect to 'http://0.0.0.0:8000/api/accounts/register/' because it violates the following Content Security Policy directive: "default-src 'self'". Note that 'connect-src' was not explicitly set, so 'default-src' is used as a fallback.

To be clear, I am not suggesting that the CSP should be changed. Rather, my point is that `0.0.0.0` is not a valid host and shouldn't be used.

In a previous review of this PR, I made this suggestion: https://github.com/Comp-490-SeniorProject/site/pull/63#discussion_r836764042. Notice that it includes a diff, which is me suggesting the exact change you should make to the code. If you follow that suggestion exactly (so it won't use any `baseUrl` at all), I can confirm it will work.

File: web/settings.py (hunk 2 above)
Comment (MarkKoz) [reply web/settings.py:84]: You still need to change this back to `SessionAuthentication`. When I was testing your PR on my machine, I changed it back (together with removing the `0.0.0.0` base URL) and the register request still worked perfectly fine.

File: web/frontend/angular/src/app/main-pages/register/register.component.ts [hunk 12]
```diff
@@ -36,7 +38,20 @@ export class RegisterComponent implements OnInit {
...
+                let input = document.getElementsByTagName("input")
+                for (let i = 0; i < input.length - 1; i++) {
+                    input[i].setAttribute("class", "error form-control")
+                }
```
[the remark this answers is not in the document]
Comment (MarkKoz) [reply web/frontend/angular/src/app/main-pages/register/register.component.ts:74]: This does not seem to be addressed. In the following screenshot, notice how all fields are red despite only the password fields being problematic. The username and e-mail are fine, so why are they red?

bild

File: web/frontend/angular/src/app/main-pages/register/register.component.ts (hunk 7 above)
Comment (MarkKoz) [reply web/frontend/angular/src/app/main-pages/register/register.component.ts:66]: This clears some errors, but not all of them.

bild

I initially had two errors: the password being short, and the passwords not matching. The short password error disappeared, but as you can see, the passwords not matching error remains.

File: web/frontend/angular/src/app/main-pages/register/register.component.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (MarkKoz) [reply web/frontend/angular/src/app/main-pages/register/register.component.ts:74]: Another issue is that the error for the username is not in red, and it does not make the field red. It should be stylistically consistent with other errors and become red.

bild

**Review by MarkKoz — CHANGES_REQUESTED** [review]

**Review by MarkKoz — APPROVED** [review]
Good work. Everything seems to be addressed now.
